sphinx/domains/c/_symbol.py

```diff
@@ -153,25 +178,28 @@ def _add_function_params(self) -> None:
             Symbol.debug_indent -= 1
 
     def remove(self) -> None:
-        if self.parent is None:
-            return
-        assert self in self.parent._children
-        self.parent._children.remove(self)
-        self.parent = None
+        if self.parent:
+            self.parent._remove_child(self)
+            self.parent = None
 
     def clear_doc(self, docname: str) -> None:
-        for sChild in self._children:
-            sChild.clear_doc(docname)
-            if sChild.declaration and sChild.docname == docname:
-                sChild.declaration = None
-                sChild.docname = None
-                sChild.line = None
-                if sChild.siblingAbove is not None:
-                    sChild.siblingAbove.siblingBelow = sChild.siblingBelow
-                if sChild.siblingBelow is not None:
-                    sChild.siblingBelow.siblingAbove = sChild.siblingAbove
-                sChild.siblingAbove = None
-                sChild.siblingBelow = None
+        if docname not in self._childrenByDocname:
+            for child in self._children:
+                child.clear_doc(docname)
+            return
+        for sChild in self._childrenByDocname[docname].values():
```

Comment:
Since `childrenByDocname[docname]` does not change during the iteration, I would suggest:

```python
children = self._childrenByDocname.pop(docname)
for child in children:
    # do the stuff
```

Reply:
Good suggestion, thanks.